evm/src/main/java/org/hyperledger/besu/evm/operation/BlockHashOperation.java

```diff
@@ -29,45 +32,93 @@
+    this(gasCalculator, BLOCK_HASH_LOOKUP);
+  }
+
+  /**
+   * Instantiates a new Block hash operation.
+   *
+   * @param gasCalculator the gas calculator
+   * @param blockHashRetrievalStrategy whether read from state (EIP-2935)
+   */
+  public BlockHashOperation(
+      final GasCalculator gasCalculator,
+      final BlockHashRetrievalStrategy blockHashRetrievalStrategy) {
     super(0x40, "BLOCKHASH", 1, 1, gasCalculator, gasCalculator.getBlockHashOperationGasCost());
+    this.blockHashRetrievalStrategy = blockHashRetrievalStrategy;
   }
 
   @Override
-  public Operation.OperationResult executeFixedCostOperation(
-      final MessageFrame frame, final EVM evm) {
+  public OperationResult executeFixedCostOperation(final MessageFrame frame, final EVM evm) {
     final Bytes blockArg = frame.popStackItem().trimLeadingZeros();
 
-    // Short-circuit if value is unreasonably large
-    if (blockArg.size() > 8) {
+    if (blockArg.size() > MAX_BLOCK_ARG_SIZE) {
       frame.pushStackItem(UInt256.ZERO);
       return successResponse;
     }
 
     final long soughtBlock = blockArg.toLong();
-    final BlockValues blockValues = frame.getBlockValues();
-    final long currentBlockNumber = blockValues.getNumber();
-    final long mostRecentBlockNumber = currentBlockNumber - 1;
-
-    // If the current block is the genesis block or the sought block is
-    // not within the last 256 completed blocks, zero is returned.
-    if (currentBlockNumber == 0
-        || soughtBlock < (mostRecentBlockNumber - MAX_RELATIVE_BLOCK)
-        || soughtBlock > mostRecentBlockNumber) {
-      frame.pushStackItem(Bytes32.ZERO);
+    final long currentBlockNumber = frame.getBlockValues().getNumber();
+
+    if (!isBlockWithinLast256Blocks(soughtBlock, currentBlockNumber)) {
+      frame.pushStackItem(UInt256.ZERO);
     } else {
-      final Function<Long, Hash> blockHashLookup = frame.getBlockHashLookup();
-      final Hash blockHash = blockHashLookup.apply(soughtBlock);
-      frame.pushStackItem(blockHash);
+      frame.pushStackItem(getBlockHash(frame, soughtBlock));
     }
 
     return successResponse;
   }
+
+  private boolean isBlockWithinLast256Blocks(
+      final long soughtBlock, final long currentBlockNumber) {
+    return soughtBlock >= Math.max(currentBlockNumber - BLOCKHASH_OLD_WINDOW, 0)
+        && soughtBlock < currentBlockNumber;
+  }
+
+  private Bytes32 getBlockHash(final MessageFrame frame, final long soughtBlock) {
+    if (blockHashRetrievalStrategy == STATE_READ) {
+      return readBlockHashFromState(frame, soughtBlock);
+    } else {
+      return lookupBlockHash(frame, soughtBlock);
+    }
+  }
+
+  private Bytes32 readBlockHashFromState(final MessageFrame frame, final long soughtBlock) {
+    Hash blockHash =
+        Hash.wrap(
+            frame
+                .getWorldUpdater()
+                .get(HISTORY_STORAGE_ADDRESS)
+                .getStorageValue(UInt256.valueOf(soughtBlock % BLOCKHASH_OLD_WINDOW)));
```

Comment:
I think we should be doing `soughtBlock % HISTORY_SERVE_WINDOW` not `soughtBlock % BLOCKHASH_OLD_WINDOW`

Reply:
I think jason is right here, we may need to add a test where the block being sought is > BLOCKHASH_OLD_WINDOW but < HISTORY_SERVE_WINDOW

Reply:
The last information I had was that this EIP was not supposed to serve more than 256 block hashes with the block hash operation, but this may have changed and needs to be verified. @Gabriel-Trintinalia

Reply:
`BLOCKHASH_OLD_WINDOW` is the serving window pre-prague.  `HISTORY_SERVE_WINDOW`  is the new serving window. There was an error in that part. Fixed now